src/rust/engine/logging/src/logger.rs

```diff
@@ -29,31 +29,65 @@ lazy_static! {
 
 pub struct PantsLogger {
   log_file: Mutex<Option<File>>,
+  global_level: Mutex<RefCell<LevelFilter>>,
   use_color: AtomicBool,
   show_rust_3rdparty_logs: AtomicBool,
   stderr_handlers: Mutex<HashMap<Uuid, StdioHandler>>,
+  show_target: AtomicBool,
+  log_level_filters: Mutex<RefCell<HashMap<String, log::LevelFilter>>>,
 }
 
 impl PantsLogger {
   pub fn new() -> PantsLogger {
     PantsLogger {
       log_file: Mutex::new(None),
+      global_level: Mutex::new(RefCell::new(LevelFilter::Off)),
       show_rust_3rdparty_logs: AtomicBool::new(true),
       use_color: AtomicBool::new(false),
       stderr_handlers: Mutex::new(HashMap::new()),
+      show_target: AtomicBool::new(false),
+      log_level_filters: Mutex::new(RefCell::new(HashMap::new())),
     }
   }
 
-  pub fn init(max_level: u64, show_rust_3rdparty_logs: bool, use_color: bool) {
+  pub fn init(
+    max_level: u64,
+    show_rust_3rdparty_logs: bool,
+    use_color: bool,
+    show_target: bool,
+    log_levels_by_target: HashMap<String, u64>,
+  ) {
+    let log_levels_by_target = log_levels_by_target
+      .iter()
+      .map(|(k, v)| {
+        let python_level: PythonLogLevel = (*v).try_into().unwrap_or_else(|e| {
+          panic!("Unrecognized log level from python: {}: {}", v, e);
+        });
+        let level: log::LevelFilter = python_level.into();
+        (k.clone(), level)
+      })
+      .collect::<HashMap<_, _>>();
+
     let max_python_level: Result<PythonLogLevel, _> = max_level.try_into();
     match max_python_level {
       Ok(python_level) => {
         let level: LevelFilter = python_level.into();
-        set_max_level(level);
+        // TODO this should be whatever the most verbose log level specified in log_domain_levels -
+        // but I'm not sure if it's actually much of a gain over just setting this to Trace.
+        set_max_level(LevelFilter::Trace);
```

Comment:
Does this mean we would be formatting the strings for log messages only to discard them?

Reply:
@tdyas with this change we would execute format strings in Rust `log!` macros (and `debug!`, etc.), but we wouldn't generate the final format string in `PantsLogger::log` since that only gets executed if `PantsLogger::enabled` is true. If the performance hit from doing too much trace-level string formatting is too bad, we could probably hook into whatever system the log macros are using to know whether they need to print, and make that aware of this option too.